File: Algorithm_LevelOne/Rotting_Oranges/Rotting_Oranges.cc
// 994. Rotting Oranges C++
#include <vector>
#include <queue>

class Solution {
public:
    int orangesRotting(std::vector<std::vector<int>>& grid) {
        if (grid.empty()) return 0;

        std::queue<std::pair<int, int>> q;
        int fresh_count = 0;
        int rows = grid.size();
        int cols = grid[0].size();

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == 2) q.push({i, j});
                else if (grid[i][j] == 1) fresh_count++;
            }
        }

        int minute = 0;
        std::vector<std::pair<int, int>> dirs = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        while (fresh_count != 0 && !q.empty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                int row = q.front().first;
                int col = q.front().second;
                q.pop();
                for (auto x: dirs) {
                    int neighbor_row = row + x.first;
                    int neighbor_col = col + x.second;
                    if (neighbor_row >= 0 && neighbor_row < rows && neighbor_col >= 0 && neighbor_col < cols && grid[neighbor_row][neighbor_col] == 1) {
                        grid[neighbor_row][neighbor_col] = 2;
                        fresh_count--;
                        q.push({neighbor_row, neighbor_col});
                    }
                }
            }
            minute++;
        }
        return fresh_count == 0 ? minute : -1;
    }
};

File: Algorithm_LevelOne/Rotting_Oranges/Rotting_Oranges.py
# 994. Rotting Oranges Python
from collections import deque

class Solution(object):
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        queue = deque()

        fresh_count = 0
        rows, cols = len(grid), len(grid[0])
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    queue.append((r, c))
                elif grid[r][c] == 1:
                    fresh_count += 1
        queue.append((-1, -1))

        minute = -1
        directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]
        while queue:
            row, col = queue.popleft()
            if row == -1:
                minute += 1
                if queue:
                    queue.append((-1, -1))
            else:
                for d in directions:
                    neighbor_row, neighbor_col = row + d[0], col + d[1]
                    if rows > neighbor_row >= 0 and cols > neighbor_col >= 0:
                        if grid[neighbor_row][neighbor_col] == 1:
                            grid[neighbor_row][neighbor_col] = 2
                            fresh_count -= 1
                            queue.append((neighbor_row, neighbor_col))
        return minute if fresh_count == 0 else -1
